playground/attention_with_rpe_template.py

```diff
@@ -0,0 +1,221 @@
+    B = tkl.sym.B
+    M = tkl.sym.M
+    N = tkl.sym.N
+    K1 = tkl.sym.K1
+    K2 = tkl.sym.K2
+    # Workgroup tile sizes
+    BLOCK_B = tkl.sym.BLOCK_B
+    BLOCK_M = tkl.sym.BLOCK_M
+    BLOCK_N = tkl.sym.BLOCK_N
+    BLOCK_K2 = tkl.sym.BLOCK_K2
+    # Address space (for GPU, shared(1) or global(0))
+    ADDRESS_SPACE = tkl.sym.ADDRESS_SPACE
+    # Other hyperparameters
+    LOAD_ELEMS_PER_THREAD_QK = index_symbol("LOAD_ELEMS_PER_THREAD_QK")
+    LOAD_ELEMS_PER_THREAD_PV = index_symbol("LOAD_ELEMS_PER_THREAD_PV")
+    STORE_ELEMS_PER_THREAD = tkl.sym.STORE_ELEMS_PER_THREAD
+
+    # Expose user-constraints
+    constraints: list[tkw.Constraint] = [
+        tkw.WorkgroupConstraint(M, BLOCK_M, 0)
+    ]
+    constraints += [tkw.WorkgroupConstraint(N, BLOCK_N, 1)]
+    constraints += [tkw.WorkgroupConstraint(B, BLOCK_B, 2)]
+    constraints += [tkw.TilingConstraint(K2, BLOCK_K2)]
+    constraints += [tkw.WaveConstraint(M, BLOCK_M / 4)]
+    constraints += [tkw.WaveConstraint(N, BLOCK_N / 1)]
+
+    if mfma_variant[1] == MMAType.F32_16x16x16_F16:
+        Mvec = 16
+        Nvec = 16
+    if mfma_variant[1] == MMAType.F32_32x32x8_F16:
+        Mvec = 32
+        Nvec = 32
+
+    constraints += [
+        tkw.HardwareConstraint(
+            threads_per_wave=64,
+            waves_per_block=(4, 1, 1),
+            mma_type=mfma_variant[1],
+            vector_shapes={
+                B: 0,
+                M: Mvec,
+                N: Nvec
+            },
+        )
+    ]
+
+    if dynamic_dims:
+        constraints += [tkw.Assumption(K2 > BLOCK_K2 * 4)]
+
+    # yapf: disable
+    i, j, k = [tkw.IndexMapping.iterator(i) for i in range(3)]
+    mapping = tkw.IndexMapping(
+        num_iterators=3,
+        inputs={ B: i, N: j, M: k },
+        outputs={ B: i, M: k, N: j })
+
+    use_t5_rpe = max_context_length is not None
+    if use_t5_rpe:
+        d0, d1 = [ tkw.IndexMapping.dynamic_val(i) for i in range(2)]
+        clip_upper = sympy.Piecewise(
+            (d0 - d1, d0 - d1 <= max_context_length),
+            (max_context_length, True)
+        )
+        clip_lower_and_upper = sympy.Piecewise(
+            (clip_upper, d0 - d1 >= 0),
+            (0, True)
+        )
+        offset_mapping = tkw.IndexMapping(
+            num_iterators=2,
+            inputs={
+                M: i,
+                K2: clip_lower_and_upper
+            },
+            outputs={ M: i, K2: j },
+            dynamic_val_mappings=({ M: i, K2: j }, { M: i, K2: j })
+        )
+
+        rpe_layout = tkl.MemoryLayout(shape=[max_context_length, ])
+    else:
+        rpe_layout = tkl.MemoryLayout(shape=[1, ])
+    # yapf: enable
+
+    @tkw.wave(constraints)
+    def base_attention(
+        q: tkl.Memory[B, M, K1, GLOBAL_ADDRESS_SPACE, tkl.f16],
+        k: tkl.Memory[B, K2, K1, ADDRESS_SPACE, tkl.f16],
+        v: tkl.Memory[B, N, K2, ADDRESS_SPACE, tkl.f16],
+        # TODO: if not use_t5_rpe, this will DCE; atm DCE on blockargs crashes.
```

Comment:
Since this is going to be t5_rpe template can you remove the use_t5_rpe in the entire template?

Reply:
ok, but note that it was by using this template and setting use_t5_rpe to false that I realized the vanilla version was incorrect on 32^n (due to no padding/masking support).
This is what prompted me to go to 128^n after looking at the WaveConstraints.